File: Embarcadero/XE6/CPP/CapitalIAP/Answered.cpp
//---------------------------------------------------------------------------

#include <fmx.h>
#pragma hdrstop

#include "Answered.h"
#include "Main.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma resource "*.fmx"

TAnsweredForm *AnsweredForm = NULL;

//---------------------------------------------------------------------------
void CreateAnswer(const TForm *Form)
{
	if(AnsweredForm == nullptr) {
		AnsweredForm = new TAnsweredForm(const_cast<TForm*>(Form));
	}
}
//---------------------------------------------------------------------------
void __fastcall TAnsweredForm::Prepare(TQuizForm *QuizForm,
	bool Correct, const String State, const String Answer)
{
	const String cCorrect = "That is correct!";
	const String cWrong = "Oh, wow, sorry";
	const String cAnswer = "%s is the capital of %s.";
	FQuizForm = QuizForm;
	if(Correct) {
		lTitle->Text = cCorrect;
	}
	else {
		lTitle->Text = cWrong;
	}
	iYes->Visible = Correct;
	iNo->Visible = !Correct;

	lAnswer->Text = Format(cAnswer, ARRAYOFCONST((Answer, State)));
}
//---------------------------------------------------------------------------
void ShowAnswer(const TForm * Form, bool Correct, const String State,
	const String Answer)
{
	CreateAnswer(Form);
	AnsweredForm->Prepare((TQuizForm*)Form, Correct, State, Answer);
	AnsweredForm->Show();
}
//---------------------------------------------------------------------------
__fastcall TAnsweredForm::TAnsweredForm(TComponent* Owner)
	: TForm(Owner)
{
}
//---------------------------------------------------------------------------
void __fastcall TAnsweredForm::FormActivate(TObject *Sender)
{
	Log::d("Setting ad parent to answered form");
	MainForm->TakeAdvertFromMainForm(this);
}
//---------------------------------------------------------------------------

void __fastcall TAnsweredForm::FormDeactivate(TObject *Sender)
{
	Log::d("etting ad parent back to main form");
	MainForm->PlaceAdvertOnMainForm();
}
//---------------------------------------------------------------------------

void __fastcall TAnsweredForm::bHomeClick(TObject *Sender)
{
	Close();
  	FQuizForm->GoHome();
}
//---------------------------------------------------------------------------

void __fastcall TAnsweredForm::bNextClick(TObject *Sender)
{
	Close();
	FQuizForm->ReviewedAnswer();
}
//---------------------------------------------------------------------------

File: Embarcadero/XE7/ObjectPascal/CapitalIAP/Answered.pas
unit Answered;

interface

uses
  System.SysUtils, System.Types, System.UITypes, System.Rtti, System.Classes,
  System.Variants, FMX.Types, FMX.Controls, FMX.Forms, FMX.Dialogs,
  FMX.StdCtrls, FMX.Objects, Quiz, FMX.Layouts;

type
  TAnsweredForm = class(TForm)
    ToolBar1: TToolBar;
    lTitle: TLabel;
    lAnswer: TLabel;
    iNo: TImage;
    Button1: TButton;
    bHome: TButton;
    iYes: TImage;
    bNext: TButton;
    Layout1: TLayout;
    procedure bHomeClick(Sender: TObject);
    procedure bNextClick(Sender: TObject);
    procedure FormActivate(Sender: TObject);
    procedure FormDeactivate(Sender: TObject);
  protected
    FQuizForm: TQuizForm;
  public
    procedure Prepare(const QuizForm: TQuizForm; Correct: Boolean; const State, Answer: string );
  end;

var
  AnsweredForm: TAnsweredForm = nil;

procedure CreateAnswer(const Form: TForm);
procedure ShowAnswer(const Form: TForm; Correct: Boolean; const State,
  Answer: string);

implementation

{$R *.fmx}

uses Main;

procedure CreateAnswer(const Form: TForm);
begin
  if not Assigned(AnsweredForm) then
    AnsweredForm := TAnsweredForm.Create(Form);
end;

procedure ShowAnswer(const Form: TForm; Correct: Boolean; const State,
  Answer: string);
begin
  CreateAnswer(Form);

  AnsweredForm.Prepare(TQuizForm(Form), Correct, State, Answer);
  AnsweredForm.Show;
end;

{ TAnsweredForm }

procedure TAnsweredForm.bHomeClick(Sender: TObject);
begin
  Close;
  FQuizForm.GoHome;
end;

procedure TAnsweredForm.bNextClick(Sender: TObject);
begin
  Close;
  FQuizForm.ReviewedAnswer;
end;

procedure TAnsweredForm.FormActivate(Sender: TObject);
begin
  Log.d('Setting ad parent to answered form');
  MainForm.TakeAdvertFromMainForm(Self);
end;

procedure TAnsweredForm.FormDeactivate(Sender: TObject);
begin
  Log.d('Setting ad parent back to main form');
  MainForm.PlaceAdvertOnMainForm;
end;

procedure TAnsweredForm.Prepare(const QuizForm: TQuizForm; Correct: Boolean; const State, Answer: string);
const
  cCorrect = 'That is correct!';
  cWrong = 'Oh, wow, sorry';
  cAnswer = '%s is the capital of %s.';
begin
  FQuizForm := QuizForm;
  if Correct then
    lTitle.Text := cCorrect
  else
    lTitle.Text := cWrong;

  iYes.Visible := Correct;
  iNo.Visible := not Correct;

  lAnswer.Text := Format(cAnswer, [Answer, State]);
end;
end.
